File: c++/approx_pi_finder.cc
#include <iostream>
#include <vector>
#include <cmath>
#include <utils.h>
#include <point.h>
#include <approx_pi_finder.h>

const double EstimatePi::ApproxPiFinder::RADIUS() const
{
    return 1000.0;
}

void EstimatePi::ApproxPiFinder::assignPoints()
{
    for(std::vector<Point>::iterator it = points.begin(); it != points.end(); ++it)
        *it = Point(randomTill(RADIUS()), randomTill(RADIUS()));
}

int EstimatePi::ApproxPiFinder::totalPointsInsideCircle() const
{
    int count = 0;
    for(std::vector<Point>::const_iterator it = points.begin(); it != points.end(); ++it)
        if (it->distance() < RADIUS())
            count++;
    return count;
}

EstimatePi::ApproxPiFinder::ApproxPiFinder(int _sample_size)
{
    sample_size = _sample_size;
    points = std::vector<Point>(sample_size);
}

double EstimatePi::ApproxPiFinder::approxPi() const
{
    return (4.0 * totalPointsInsideCircle()) / sample_size;
}

double EstimatePi::ApproxPiFinder::error() const
{
    return (approxPi() - actualPi()) * 100.0 / actualPi();
}

void EstimatePi::ApproxPiFinder::evaluate()
{
    assignPoints();
}

void EstimatePi::ApproxPiFinder::outputToConsole() const
{
    std::cout << std::endl;
    std::cout << "Approx pi  : " << approxPi() << std::endl;
    std::cout << "Actual pi  : " << actualPi() << std::endl;
    std::cout << "Error      : " << error() << "%" << std::endl;
}

File: java/ApproxPiFinder.java
import java.util.ArrayList;

class ApproxPiFinder
{
    private static double RADIUS = 1000.0;
    private int sampleSize;

    private ArrayList<Point> points;

    private void assignPoints() {
        for (int i = 0; i < sampleSize; i++)
            points.add(new Point(Utils.randomTill(RADIUS), Utils.randomTill(RADIUS)));
    }

    private int totalPointsInsideCircle() {
        int count = 0;
        for(Point point : points)
            if (point.distance() < RADIUS)
                count++;
        return count;
    }

    public ApproxPiFinder(int sampleSize) {
        this.sampleSize = sampleSize;
        this.points = new ArrayList<Point>(this.sampleSize);
    }

    public double approxPi() {
        return (4.0 * totalPointsInsideCircle()) / sampleSize;
    }

    public double error() {
        return (approxPi() - Utils.actualPi()) * 100.0 / Utils.actualPi();
    }

    public void evaluate() {
        assignPoints();
    }

    public void outputToConsole() {
        System.out.println("");
        System.out.println("Approx pi  : " + approxPi());
        System.out.println("Actual pi  : " + Utils.actualPi());
        System.out.println("Error      : " + error() + "%");
    }
}

